File: scripts/scrape.js
const cheerio = require('cheerio');
const fs = require('fs');
const path = require('path');

const fetch = (...args) =>
  import('node-fetch').then(({ default: fetch }) => fetch(...args));

// 日付
// UTC を取得
const now = new Date();

// JST にずらす
const jst = new Date(now.getTime() + (9 * 60 * 60 * 1000));

// JST で md を作る
const md = `${String(jst.getMonth() + 1).padStart(2, '0')}${String(jst.getDate()).padStart(2, '0')}`;

// URL
const url = `http://www.data.jma.go.jp/obd/stats/data/mdrr/rank_daily/data${md}.html`;

// 地点マスタをロード
const amedasPath = path.resolve(__dirname, '../public/amedas.json');
const amedas = JSON.parse(fs.readFileSync(amedasPath, 'utf8'));

console.log(`🌐 ${url} を取得中...`);

fetch(url)
  .then(res => res.text())
  .then(html => {
    const $ = cheerio.load(html);

    const clean = (str) => str.replace(/[\]\)\（\(\s]+$/, '').trim();

    const joinAmedas = (pref, city) => {
      return amedas.find(a =>
        a.group.includes(pref) &&
        (city.includes(a.name) || a.name.includes(city))
      );
    };

    // === High ===
    const rowsHigh = $('div#main table.data2_s:nth-of-type(1) tr.mtx');
    const resultHigh = [];
    let prevRankHigh = null;

    rowsHigh.each((i, el) => {
      const tds = $(el).find('td');
      if (!$(tds[0]).text().trim()) return;

      let rank = $(tds[0]).text().trim();
      if (rank === '〃') rank = prevRankHigh;
      prevRankHigh = rank;

      const pref = $(tds[1]).text().trim();
      const city = $(tds[3]).text().trim().replace(/\（.*\）/, '').replace('*', '');

      const match = joinAmedas(pref, city);

      resultHigh.push({
        rank,
        pref,
        city,
        temp: clean($(tds[4]).text()),
        time: clean($(tds[5]).text()),
        memo: $(tds[tds.length - 1]).text().trim(),
        code: match ? match.code : null,
        group: match ? match.group : null,
        kana: match ? match.kana : null,
        lat: match ? match.lat : null,
        lon: match ? match.lon : null,
      });
    });

    // === Low ===
    const rowsLow = $('div#main table.data2_s:nth-of-type(2) tr.mtx');
    const resultLow = [];
    let prevRankLow = null;

    rowsLow.each((i, el) => {
      const tds = $(el).find('td');
      if (!$(tds[0]).text().trim()) return;

      let rank = $(tds[0]).text().trim();
      if (rank === '〃') rank = prevRankLow;
      prevRankLow = rank;

      const pref = $(tds[1]).text().trim();
      const city = $(tds[3]).text().trim().replace(/\（.*\）/, '').replace('*', '');

      const match = joinAmedas(pref, city);

      resultLow.push({
        rank,
        pref,
        city,
        temp: clean($(tds[4]).text()),
        time: clean($(tds[5]).text()),
        memo: $(tds[tds.length - 1]).text().trim(),
        code: match ? match.code : null,
        group: match ? match.group : null,
        kana: match ? match.kana : null,
        lat: match ? match.lat : null,
        lon: match ? match.lon : null,
      });
    });

    console.log(`✅ High: ${resultHigh.length}件, Low: ${resultLow.length}件`);

    // 取得した HTML から見出しを取る
    const nowText = $('div#main h1').text().trim();

    // 正規表現で月・日・時・分を取る
    const match = nowText.match(/（(\d+)月(\d+)日）(\d+)時(\d+)分/);

    let ymdhm = '';
    if (match) {
      const [, m, d, h, min] = match;
      const y = now.getFullYear();
      ymdhm = `${y}${String(m).padStart(2, '0')}${String(d).padStart(2, '0')}-${String(h).padStart(2, '0')}${String(min).padStart(2, '0')}`;
      console.log(`📅 公式基準: ${ymdhm}`);
    } else {
      console.log('⚠️ 日時抽出失敗 → fallback to local time');
      ymdhm = `${now.getFullYear()}${String(now.getMonth() + 1).padStart(2, '0')}${String(now.getDate()).padStart(2, '0')}-${String(now.getHours()).padStart(2, '0')}${String(now.getMinutes()).padStart(2, '0')}`;
    }

    // 保存先
    const highDir = path.resolve(__dirname, '../www/428ch/hightemp-map/data/high');
    const lowDir = path.resolve(__dirname, '../www/428ch/hightemp-map/data/low');

    fs.mkdirSync(highDir, { recursive: true });
    fs.mkdirSync(lowDir, { recursive: true });

    fs.writeFileSync(path.join(highDir, 'latest.json'), JSON.stringify(resultHigh, null, 2));
    fs.writeFileSync(path.join(lowDir, 'latest.json'), JSON.stringify(resultLow, null, 2));
    fs.writeFileSync(path.join(highDir, `${ymdhm}.json`), JSON.stringify(resultHigh, null, 2));
    fs.writeFileSync(path.join(lowDir, `${ymdhm}.json`), JSON.stringify(resultLow, null, 2));

    console.log(`🗂️ 保存完了: latest.json & ${ymdhm}.json`);
  })
  .catch(err => {
    console.error('❌ 取得失敗:', err);
  });

File: scripts/scrape.py
# -*- coding: utf-8 -*-
import urllib2
import sys
import json
import os
import datetime
import re
import codecs
from BeautifulSoup import BeautifulSoup

# JST 現在日時
now = datetime.datetime.utcnow() + datetime.timedelta(hours=9)
md = '%02d%02d' % (now.month, now.day)

# URL
url = 'http://www.data.jma.go.jp/obd/stats/data/mdrr/rank_daily/data%s.html' % md

print('%s を取得中...' % url)

# 地点マスタ
amedas_path = os.path.join(os.path.dirname(__file__), 'amedas.json')
with open(amedas_path, 'r') as f:
    amedas = json.load(f)

response = urllib2.urlopen(url)
html = response.read()
soup = BeautifulSoup(html)

def clean(text):
    return re.sub(u'[\\]\\)\\（\\(\\s]+$', '', text.strip())

def join_amedas(pref, city):
    for a in amedas:
        if pref in a['group'] and (a['name'] in city or city in a['name']):
            return a
    return None

# === High ===
tables = soup.findAll('table', {'class': 'data2_s'})
rows_high = tables[0].findAll('tr', {'class': 'mtx'})
result_high = []
prev_rank_high = None

for row in rows_high:
    tds = row.findAll('td')
    if not tds or not tds[0].text.strip():
        continue
    rank = tds[0].text.strip()
    if rank == u'〃':
        rank = prev_rank_high
    prev_rank_high = rank
    pref = tds[1].text.strip()
    city = re.sub(u'（.*）', '', tds[3].text.strip()).replace('*', '')
    match = join_amedas(pref, city)
    result_high.append({
        'rank': rank,
        'pref': pref,
        'city': city,
        'temp': clean(tds[4].text.strip()),
        'time': clean(tds[5].text.strip()),
        'memo': tds[-1].text.strip(),
        'code': match['code'] if match else None,
        'group': match['group'] if match else None,
        'kana': match['kana'] if match else None,
        'lat': match['lat'] if match else None,
        'lon': match['lon'] if match else None
    })

print('✅ High: %d件' % len(result_high))

# === Low ===
rows_low = tables[1].findAll('tr', {'class': 'mtx'})
result_low = []
prev_rank_low = None

for row in rows_low:
    tds = row.findAll('td')
    if not tds or not tds[0].text.strip():
        continue
    rank = tds[0].text.strip()
    if rank == u'〃':
        rank = prev_rank_low
    prev_rank_low = rank
    pref = tds[1].text.strip()
    city = re.sub(u'（.*）', '', tds[3].text.strip()).replace('*', '')
    match = join_amedas(pref, city)
    result_low.append({
        'rank': rank,
        'pref': pref,
        'city': city,
        'temp': clean(tds[4].text.strip()),
        'time': clean(tds[5].text.strip()),
        'memo': tds[-1].text.strip(),
        'code': match['code'] if match else None,
        'group': match['group'] if match else None,
        'kana': match['kana'] if match else None,
        'lat': match['lat'] if match else None,
        'lon': match['lon'] if match else None
    })

print('✅ Low: %d件' % len(result_low))

# 公式基準日時
now_text = soup.find('h1').text.strip()
m = re.search(u'（(\d+)月(\d+)日）(\d+)時(\d+)分', now_text)
if m:
    ymdhm = '%04d%02d%02d-%02d%02d' % (now.year, int(m.group(1)), int(m.group(2)), int(m.group(3)), int(m.group(4)))
    ymd = '%02d%02d' % (int(m.group(1)), int(m.group(2)))
    hm = '%02d%02d' % (int(m.group(3)), int(m.group(4)))
    print('📅 公式基準: %s' % ymdhm)
else:
    ymdhm = now.strftime('%Y%m%d-%H%M')
    ymd = now.strftime('%m%d')
    hm = now.strftime('%H%M')
    print('⚠️ 日時抽出失敗 → fallback: %s' % ymdhm)

# 保存先
base_dir = '/home/shibuya/www/428ch/hightemp-map/data'
high_dir = os.path.join(base_dir, 'high')
low_dir = os.path.join(base_dir, 'low')

if not os.path.exists(high_dir):
    os.makedirs(high_dir)
if not os.path.exists(low_dir):
    os.makedirs(low_dir)

with codecs.open(os.path.join(high_dir, 'latest.json'), 'w', 'utf-8') as f:
    json.dump(result_high, f, ensure_ascii=False, indent=2)
with codecs.open(os.path.join(low_dir, 'latest.json'), 'w', 'utf-8') as f:
    json.dump(result_low, f, ensure_ascii=False, indent=2)

with codecs.open(os.path.join(high_dir, '%s.json' % ymdhm), 'w', 'utf-8') as f:
    json.dump(result_high, f, ensure_ascii=False, indent=2)
with codecs.open(os.path.join(low_dir, '%s.json' % ymdhm), 'w', 'utf-8') as f:
    json.dump(result_low, f, ensure_ascii=False, indent=2)

print('🗂️ 保存完了: latest.json & %s.json' % ymdhm)

# === 履歴ファイル更新 ===
def update_history(type_name, ymd, hm):
    path = os.path.join(base_dir, 'history-%s.json' % type_name)
    if os.path.exists(path):
        with codecs.open(path, 'r', 'utf-8') as f:
            history = json.load(f)
    else:
        history = {'grouped': {}, 'recent': []}

    if ymd not in history['grouped']:
        history['grouped'][ymd] = []
    if hm not in history['grouped'][ymd]:
        history['grouped'][ymd].append(hm)

    key = '%s-%s' % (ymd, hm)
    if key not in history['recent']:
        history['recent'].insert(0, key)
        history['recent'] = history['recent'][:10]

    with codecs.open(path, 'w', 'utf-8') as f:
        json.dump(history, f, ensure_ascii=False, indent=2)
    print('✅ Updated history-%s.json' % type_name)

update_history('high', ymd, hm)
update_history('low', ymd, hm)

print('🎉 すべて完了')
